File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/Permutations.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class Permutations {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = permutations(a);
        for (var r : res) {
          for (int e : r) {
            System.out.printf("%d ", e);
          }
          System.out.println();
        }
      }
    }
  }

  // O(n x n!) / O(n x n!)
  public static List<List<Integer>> permutations(int[] a) {
    List<List<Integer>> res = new ArrayList<>();

    Queue<List<Integer>> q = new ArrayDeque<>();
    q.add(new LinkedList<>());

    for (int e : a) {
      int n = q.size();

      for (int i = 0; i < n; i++) {
        List<Integer> currPerm = q.poll();

        for (int j = 0; j <= currPerm.size(); j++) {
          List<Integer> newPerm = new LinkedList<>(currPerm);

          newPerm.add(j, e);

          if (newPerm.size() == a.length) {
            res.add(newPerm);
          } else {
            q.add(newPerm);
          }
        }
      }
    }

    return res;
  }
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/permutations.cpp
#include <algorithm>
#include <iostream>
#include <list>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

// O(n x n!) / O(n x n!)
vector<vector<int>> permutations(const vector<int> &a) {
  vector<vector<int>> res;

  queue<vector<int>> q;
  q.push(vector<int>{});

  for (int e : a) {
    int n = q.size();

    for (int i = 0; i < n; i++) {
      vector<int> curr_perm = q.front();
      q.pop();

      for (int i = 0; i <= curr_perm.size(); i++) {
        vector<int> new_perm = curr_perm;
        new_perm.insert(new_perm.begin() + i, e);

        if (new_perm.size() == a.size()) {
          res.push_back(new_perm);
        } else {
          q.push(new_perm);
        }
      }
    }
  }

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto res = permutations(a);
    for (auto r : res) {
      for (int e : r) {
        cout << e << " ";
      }
      cout << "\n";
    }
  }

  return 0;
}